In [2]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer,ENGLISH_STOP_WORDS



# Part 1 Understanding the data

In [5]:
dataset = pd.read_csv('English Dataset.csv')

# since ArticleId has no effect to classify, we ignored this column

dataset.drop('ArticleId', inplace=True, axis=1) 
dataset = dataset.reset_index(drop=True) 

# distinct categories
categories = dataset['Category'].unique()


a = dataset.to_numpy()
print(a)

cat_business = []
cat_tech = []
cat_politics = []
cat_sports = []
cat_entertainment = []


# collecting whole texts of each category
whole_category = [cat_business, cat_tech, cat_politics, cat_sports, cat_entertainment]

for i in range(len(a)):
    if a[i][1] == categories[0]:
        cat_business.append(a[i][0])
        
    elif a[i][1] == categories[1]:
        cat_tech.append(a[i][0])
        
    elif a[i][1] == categories[2]:
        cat_politics.append(a[i][0])
        
    elif a[i][1] == categories[3]:
        cat_sports.append(a[i][0])
        
    elif a[i][1] == categories[4]:
        cat_entertainment.append(a[i][0])
        
from collections import defaultdict
#         
vectorizer = CountVectorizer()

for c in range(len(whole_category)):
    
    cat_name = whole_category[c] 
    
    # for each category we create a matrix
    X = vectorizer.fit_transform(cat_name) 
    matrix = X.toarray()

    # collecting columns of matrix which gives words of each category         
    columns = (vectorizer.get_feature_names_out())
    
    # purpose of the summation of the columns is finding the word which has most frequency
    sumOfColumns = np.sum(matrix, axis=0)

    # keeping words and number of words in dict_word for each category
    dict_word = defaultdict(int)
    print(categories[c] , '-->', end=' ')
    
    # finding most frequent words for each category

    for i in range(50):

        max_value = np.max(sumOfColumns)

        for j in range(len(sumOfColumns)):
            if sumOfColumns[j] == max_value:
                dict_word[columns[j]] = max_value
                
        # result means index of first max summation
        result = np.where(sumOfColumns == max_value)[0][0]
        # the reason behind assinging result to zero is that not considering the same max value
        sumOfColumns[result] = 0
    print(*dict_word.items())
    
    print()



[['worldcom ex-boss launches defence lawyers defending former worldcom chief bernie ebbers against a battery of fraud charges have called a company whistleblower as their first witness.  cynthia cooper  worldcom s ex-head of internal accounting  alerted directors to irregular accounting practices at the us telecoms giant in 2002. her warnings led to the collapse of the firm following the discovery of an $11bn (£5.7bn) accounting fraud. mr ebbers has pleaded not guilty to charges of fraud and conspiracy.  prosecution lawyers have argued that mr ebbers orchestrated a series of accounting tricks at worldcom  ordering employees to hide expenses and inflate revenues to meet wall street earnings estimates. but ms cooper  who now runs her own consulting business  told a jury in new york on wednesday that external auditors arthur andersen had approved worldcom s accounting in early 2001 and 2002. she said andersen had given a  green light  to the procedures and practices used by worldcom. mr e

OUTPUT :

business --> ('the', 7133) ('to', 3306) ('of', 2864) ('in', 2821) ('and', 2161) ('said', 1100) ('is', 1072) ('that', 1052) ('for', 1045) ('it', 1011) ('on', 906) ('has', 835) ('its', 736) ('by', 718) ('with', 612) ('at', 609) ('as', 605) ('was', 596) ('from', 588) ('be', 573) ('have', 554) ('are', 538) ('us', 522) ('will', 520) ('year', 456) ('which', 404) ('mr', 393) ('but', 391) ('an', 385) ('had', 343) ('this', 322) ('would', 309) ('been', 307) ('up', 306) ('not', 302) ('more', 298) ('he', 287) ('market', 284) ('were', 282) ('also', 279) ('than', 275) ('new', 273) ('their', 265) ('firm', 261) ('growth', 257) ('company', 253) ('last', 236) ('economy', 233) ('about', 228) ('after', 221)

tech --> ('the', 7498) ('to', 4149) ('of', 3425) ('and', 3017) ('in', 2316) ('that', 1676) ('is', 1597) ('it', 1465) ('for', 1299) ('on', 1111) ('said', 1064) ('be', 1057) ('are', 994) ('as', 929) ('will', 797) ('with', 787) ('have', 731) ('by', 727) ('has', 686) ('was', 653) ('people', 647) ('they', 630) ('more', 624) ('not', 535) ('at', 533) ('but', 527) ('which', 514) ('from', 500) ('he', 484) ('can', 480) ('this', 474) ('or', 452) ('their', 422) ('up', 413) ('an', 408) ('its', 388) ('about', 371) ('one', 351) ('you', 351) ('mr', 349) ('new', 349) ('also', 348) ('were', 344) ('mobile', 343) ('than', 343) ('we', 336) ('would', 322) ('been', 312) ('could', 308) ('technology', 303)

politics --> ('the', 7957) ('to', 3913) ('of', 2840) ('and', 2559) ('in', 2159) ('said', 1445) ('he', 1410) ('for', 1237) ('that', 1195) ('on', 1186) ('is', 1167) ('be', 1080) ('mr', 1073) ('was', 1013) ('it', 998) ('would', 712) ('as', 690) ('have', 669) ('but', 668) ('not', 662) ('by', 651) ('with', 645) ('will', 642) ('are', 635) ('has', 611) ('they', 572) ('at', 564) ('his', 540) ('labour', 494) ('from', 466) ('government', 464) ('had', 462) ('an', 429) ('election', 424) ('we', 416) ('this', 410) ('blair', 395) ('were', 379) ('been', 377) ('party', 376) ('people', 372) ('there', 364) ('their', 357) ('which', 340) ('who', 330) ('also', 308) ('more', 298) ('if', 289) ('minister', 286) ('up', 284)

sport --> ('the', 6620) ('to', 3189) ('and', 2532) ('in', 2510) ('of', 1826) ('for', 1127) ('he', 1105) ('on', 1014) ('but', 992) ('is', 985) ('it', 974) ('was', 943) ('that', 863) ('have', 812) ('with', 803) ('at', 794) ('his', 762) ('we', 660) ('has', 650) ('said', 636) ('be', 614) ('will', 575) ('as', 547) ('not', 490) ('from', 481) ('after', 477) ('by', 430) ('had', 414) ('they', 414) ('their', 381) ('been', 363) ('are', 356) ('game', 356) ('an', 353) ('this', 353) ('out', 351) ('first', 350) ('year', 331) ('england', 329) ('who', 324) ('against', 312) ('time', 296) ('when', 295) ('win', 295) ('up', 294) ('two', 290) ('world', 269) ('all', 268) ('over', 267) ('there', 264)

entertainment --> ('the', 5822) ('and', 2130) ('to', 2108) ('of', 2048) ('in', 1996) ('for', 1097) ('on', 881) ('was', 818) ('it', 738) ('is', 684) ('with', 662) ('said', 594) ('he', 584) ('film', 583) ('at', 554) ('be', 502) ('that', 491) ('as', 479) ('has', 471) ('by', 460) ('his', 460) ('will', 433) ('best', 430) ('have', 384) ('who', 358) ('from', 356) ('are', 331) ('but', 326) ('year', 315) ('which', 305) ('an', 299) ('also', 277) ('been', 276) ('this', 266) ('one', 265) ('us', 264) ('had', 263) ('were', 262) ('music', 255) ('not', 251) ('they', 241) ('she', 240) ('we', 236) ('their', 235) ('new', 234) ('up', 225) ('show', 222) ('her', 215) ('first', 188) ('after', 186)

most meaningful 3 words of each category:

total words of business is 112281
Business -->  market : 284, firm : 261, growth : 257

total words of Tech is 130985
Tech --> people : 647, mobile : 343 technology : 303

total words of Politics is 123215   
Politics --> labour : 494, government : 464, election : 424
  
total words of Sports is 116030
Sports --> game : 356, first : 350, england : 329 

total words of Entertainment is 91158
Entertainment --> film : 583, best : 430, music : 255 
    
    
    
    
    
    
    
    

# Part 2: Implementing Naive Bayes

In [69]:


dataset_array = dataset.to_numpy()

np.random.shuffle(dataset_array) # shuffling dataset_array
test = dataset_array[4 * len(dataset_array) // 5:].copy()
training = dataset_array[:4 * len(dataset_array) // 5].copy()
cat_business = []
cat_tech = []
cat_politics = []
cat_sports = []
cat_entertainment = []

whole_category = [cat_business, cat_tech, cat_politics, cat_sports, cat_entertainment]

# listing lengths of total words in every categories. 
length_of_words_categorically = [0,0,0,0,0]

# collecting whole texts of each category in addition , finding total number of words of each category.


for i in range(len(dataset_array)):
    if dataset_array[i][1] == categories[0]:
        cat_business.append(dataset_array[i][0])
        length_of_words_categorically[0] += len(dataset_array[i][0].split())
    elif dataset_array[i][1] == categories[1]:
        cat_tech.append(dataset_array[i][0])
        length_of_words_categorically[1] += len(dataset_array[i][0].split())
    elif dataset_array[i][1] == categories[2]:
        cat_politics.append(dataset_array[i][0])
        length_of_words_categorically[2] += len(dataset_array[i][0].split())
    elif dataset_array[i][1] == categories[3]:
        cat_sports.append(dataset_array[i][0])
        length_of_words_categorically[3] += len(dataset_array[i][0].split())
    elif dataset_array[i][1] == categories[4]:
        cat_entertainment.append(dataset_array[i][0])
        length_of_words_categorically[4] += len(dataset_array[i][0].split())
row = dataset.shape[0] # number of texts
from collections import defaultdict
vectorizer = CountVectorizer() # CountVectorizer for unigram matrix.
vectorizer2 = CountVectorizer(ngram_range=(2, 2))# CountVectorizer for bigram matrix.
whole_matrix = vectorizer.fit_transform(training[:, 0]).toarray()
number_of_unique_words = whole_matrix.shape[1]
whole_matrix_of_bigrams = vectorizer2.fit_transform(training[:, 0]).toarray()
number_of_unique_bigrams = whole_matrix_of_bigrams.shape[1]
bag_of_unigrams = defaultdict()
bag_of_bigrams = defaultdict()
# according to unigram and bigram matrices, creating dictionaries respectively
for c in range(len(whole_category)):
    cat_name = whole_category[c]
    X = vectorizer.fit_transform(cat_name)
    unigram_matrix = X.toarray()
    words = (vectorizer.get_feature_names_out())
    bigram_matrix = vectorizer2.fit_transform(cat_name).toarray()
    bigrams = vectorizer2.get_feature_names_out()
    vocabulary = defaultdict(int) # vocabulary of unigrams for each category
    vocabulary2 = defaultdict(int) # vocabulary of bigrams for each category
    sumOfColumns = np.sum(unigram_matrix, axis=0)
    # adding existing words to unigram vocabulary
    for i in range(unigram_matrix.shape[0]):
        for j in range(unigram_matrix.shape[1]):
            if unigram_matrix[i][j] == 0:
                continue
            else:
                vocabulary[words[j]] += unigram_matrix[i][j]
    # collecting unigram vocabularies for each category
    bag_of_unigrams[categories[c]] = vocabulary
    # adding existing words to bigram vocabulary
    for i in range(bigram_matrix.shape[0]):
        for j in range(bigram_matrix.shape[1]):
            if bigram_matrix[i][j] == 0:
                continue
            else:
                vocabulary2[bigrams[j]] += bigram_matrix[i][j]
    # collecting bigram vocabularies for each category
    bag_of_bigrams[categories[c]] = vocabulary2
    
def naive_bayes(unigrams,bigrams,test):
    test = np.ndarray.tolist(test)
    import re
    global number_of_unique_words, length_of_words_categorically
    for i in range(len(test)):
        test[i] = test[i][0].lower() # converting every letter to lowercase 
        test[i] = re.sub(r'[^a-zA-Z0-9]', ' ', test[i]) # we are removing punctuations for not considering them as a word
        test[i] = test[i].split() # converting text string to string list of every word.
    predictions_unigram = []
    predictions_bigram = []
    for i in range(len(test)):
        scores_unigram = [0,0,0,0,0] # keeping scores of each category for unigram
        scores_bigram = [0,0,0,0,0] # keeping scores of each category for bigram
        for j in range(len(test[i])):
            for k in range(5):
                if test[i][j] in unigrams[categories[k]].keys():
                    scores_unigram[k] += np.log((unigrams[categories[k]][test[i][j]] +1) / (length_of_words_categorically[k] + number_of_unique_words))
                else:
                    scores_unigram[k] += np.log(1 / (length_of_words_categorically[k] + number_of_unique_words))
                if j+2 < len(test[i]):
                    bigram = test[i][j] + ' ' + test[i][j+1]
                    if bigram in bigrams[categories[k]].keys():
                        scores_bigram[k] += np.log((bigrams[categories[k]][bigram] + 1) / (length_of_words_categorically[k] + number_of_unique_words))
                    else:
                        scores_bigram[k] += np.log(1/(length_of_words_categorically[k]+number_of_unique_words))

        predictions_unigram.append(categories[scores_unigram.index(max(scores_unigram))])
        predictions_bigram.append(categories[scores_bigram.index(max(scores_bigram))])
    return predictions_unigram , predictions_bigram

def accuracy(outputs,predictions):
    correct_predictions = 0
    for i in range(len(outputs)):
        if outputs[i] == predictions[i]:
            correct_predictions +=1
    return 100 * correct_predictions / len(outputs)

predictions = naive_bayes(bag_of_unigrams,bag_of_bigrams,test)
print("Unigram: ",accuracy(test[:,1],predictions[0]))
print("Bigram: ",accuracy(test[:,1],predictions[1]))


Unigram:  99.32885906040268
Bigram:  100.0


OUTPUT:
    
Unigram:  99.66442953020135

Bigram:  100.0

OUTPUT-2:

Unigram:  98.99328859060402

Bigram:  99.66442953020135

After every running,
implementation of naive bayes for bigram, classifies better than Unigram


# Part 3 

# Listing 10 words whose presence most strongly predicts

# First Perspective of Listing 10 words whose presence most strongly predicts

We removed words if their idf is equal to 1, because meaning of idf is 1, the word exists in each text of that category so that word not too specific and strong. 
Combination of tf-idf and idf to get 10 most strongly predictions

In [70]:
vectorizer3 = TfidfVectorizer()
specific_word_for_each_category = []
for c in range(len(whole_category)):
    vectorizer3 = TfidfVectorizer()
    X = vectorizer3.fit_transform(whole_category[c])
    idfs = vectorizer3.idf_  

    wordList = vectorizer3.get_feature_names_out()

    tfidf_matrix = X.toarray()

    specific_word = []
    # to find the most frequent words 
    sumOfCol = np.sum(X.toarray(), axis=0)

    for k in range(15):  # this range should be at least the length of the list (10)
        max_val = np.max(sumOfCol)

        for i in range(len(wordList)):
            # meaning of idf is 1, the word exists in each text of that category. 
            # That's why we remove this word which is not too strong to predict the class 
            if idfs[i] == 1:
                sumOfCol[i] = 0
            if max_val != 0:
                # after finding the most frequent word, adding to the list 
                # and update the index of sumOfCol for not considering it again 
                if sumOfCol[i] == max_val:
                    specific_word.append(wordList[i])
                    sumOfCol[i] = 0
        # when the length of the list reached to 10, break the the loop
        if len(specific_word) == 10: 
            break
    specific_word_for_each_category.append(specific_word)
    print(categories[c], ": ", specific_word)



business :  ['in', 'and', 'said', 'that', 'is', 'it', 'for', 'on', 'has', 'its']
tech :  ['of', 'in', 'that', 'is', 'it', 'for', 'said', 'on', 'be', 'are']
politics :  ['of', 'and', 'he', 'said', 'mr', 'for', 'on', 'that', 'is', 'be']
sport :  ['to', 'in', 'and', 'of', 'he', 'is', 'we', 'for', 'it', 'on']
entertainment :  ['to', 'and', 'in', 'of', 'for', 'film', 'on', 'was', 'it', 'he']


Output of First Perspective of Listing 10 words whose presence most strongly predicts:

business :  ['in', 'and', 'said', 'that', 'is', 'it', 'for', 'on', 'has', 'its']
    
tech :  ['of', 'in', 'that', 'is', 'it', 'for', 'said', 'on', 'be', 'are']
    
politics :  ['of', 'and', 'he', 'said', 'mr', 'for', 'on', 'that', 'is', 'be']
    
sport :  ['to', 'in', 'and', 'of', 'he', 'is', 'we', 'for', 'it', 'on']
    
entertainment :  ['to', 'and', 'in', 'of', 'for', 'film', 'on', 'was', 'it', 'he']

# Second perspective of Listing 10 words whose presence most strongly predicts

In this part we include words even if their idf is 1

In [71]:
vectorizer3 = TfidfVectorizer()
specific_word_for_each_category = []
for c in range(len(whole_category)):
    vectorizer3 = TfidfVectorizer()
    X = vectorizer3.fit_transform(whole_category[c])
    idfs = vectorizer3.idf_  

    wordList = vectorizer3.get_feature_names_out()

    tfidf_matrix = X.toarray()

    specific_word = []

    sumOfCol = np.sum(X.toarray(), axis=0)

    for k in range(15):  # this range should be at least the length of the list (10)
        max_val = np.max(sumOfCol)

        for i in range(len(wordList)):
            
            if max_val != 0:
                if sumOfCol[i] == max_val:
                    specific_word.append(wordList[i])
                    sumOfCol[i] = 0

        if len(specific_word) == 10:
            break
    specific_word_for_each_category.append(specific_word)
    print(categories[c], ": ", specific_word)


business :  ['the', 'to', 'in', 'of', 'and', 'said', 'that', 'is', 'it', 'for']
tech :  ['the', 'to', 'of', 'and', 'in', 'that', 'is', 'it', 'for', 'said']
politics :  ['the', 'to', 'of', 'and', 'in', 'he', 'said', 'mr', 'for', 'on']
sport :  ['the', 'to', 'in', 'and', 'of', 'he', 'is', 'we', 'for', 'it']
entertainment :  ['the', 'to', 'and', 'in', 'of', 'for', 'film', 'on', 'was', 'it']


Output Of Second perspective of Listing 10 words whose presence most strongly predicts:

business :  ['the', 'to', 'in', 'of', 'and', 'said', 'that', 'is', 'it', 'for']
    
tech :  ['the', 'to', 'of', 'and', 'in', 'that', 'is', 'it', 'for', 'said']
    
politics :  ['the', 'to', 'of', 'and', 'in', 'he', 'said', 'mr', 'for', 'on']
    
sport :  ['the', 'to', 'in', 'and', 'of', 'he', 'is', 'we', 'for', 'it']
    
entertainment :  ['the', 'to', 'and', 'in', 'of', 'for', 'film', 'on', 'was', 'it']

# List the 10 words whose absence most strongly predicts 

Explaining logic of finding 10 words whose absence most strongly predicts is that:
when idf of a word has max value, then this word appears rarely in the list 
and when sum of columns of tf-idf has min value, then it shows this word appearing rarely 

In [72]:

absence_word_for_each_category = []
for c in range(len(whole_category)):
    vectorizer3 = TfidfVectorizer()
    Y = vectorizer3.fit_transform(whole_category[c])
    idfs = vectorizer3.idf_ - 1

    wordList = vectorizer3.get_feature_names_out()

    tfidf_mat = Y.toarray()

    sumOfCol = np.sum(Y.toarray(), axis=0)

    max_idf = np.max(idfs)

    min_sum = np.min(sumOfCol)

    absence_word = []

    for i in range(len(wordList)):
        if idfs[i] == max_idf:
            if sumOfCol[i] == min_sum:
                absence_word.append(wordList[i])

            if len(absence_word) == 10:
                break
    absence_word_for_each_category.append(absence_word)
    print(categories[c], ": ", absence_word)



business :  ['accessing', 'adequate', 'boundary', 'bracing', 'brighten', 'calculate', 'carefully', 'chairs', 'checked', 'circuits']
tech :  ['amazed', 'astro', 'brits', 'cant', 'coffee', 'compass', 'diamond', 'divides', 'fizzy', 'froth']
politics :  ['20p', '3rds', '3x', '5000', '50pc', '75p', '80s', 'absorb', 'adair', 'afloat']
sport :  ['aggression', 'applaud', 'avoids', 'bangs', 'boil', 'buoyed', 'buts', 'characteristic', 'chases', 'cheering']
entertainment :  ['11m', '42m', '60s', 'abating', 'admission', 'advances', 'advert', 'affection', 'airbrushes', 'amateurish']


Output of absecence list:

business :  ['accessing', 'adequate', 'boundary', 'bracing', 'brighten', 'calculate', 'carefully', 'chairs', 'checked', 'circuits']
    
tech :  ['amazed', 'astro', 'brits', 'cant', 'coffee', 'compass', 'diamond', 'divides', 'fizzy', 'froth']
    
politics :  ['20p', '3rds', '3x', '5000', '50pc', '75p', '80s', 'absorb', 'adair', 'afloat']
    
sport :  ['aggression', 'applaud', 'avoids', 'bangs', 'boil', 'buoyed', 'buts', 'characteristic', 'chases', 'cheering']
    
entertainment :  ['11m', '42m', '60s', 'abating', 'admission', 'advances', 'advert', 'affection', 'airbrushes', 'amateurish']

# Removing Stopwords

In [73]:

ENGLISH_STOP_WORDS = ENGLISH_STOP_WORDS.union(
    ["said"])  # since every category has 'said', we carried 'said' to stop_word_list
non_stop_specific_for_each_category = []
for c in range(len(whole_category)):
    vectorizer = TfidfVectorizer(stop_words=ENGLISH_STOP_WORDS)
    X = vectorizer.fit_transform(whole_category[c])
    idfs = vectorizer.idf_ - 1

    wordList = vectorizer.get_feature_names_out()

    tfidf_matrix = X.toarray()


    specific_word = []

    sumOfCol = np.sum(X.toarray(), axis=0)

    for k in range(22):  # this range should be at least the length of the list (10)
        max_val = np.max(sumOfCol)

        for i in range(len(wordList)):
            if max_val != 0:
                if sumOfCol[i] == max_val:
                    specific_word.append(wordList[i])
                    sumOfCol[i] = 0

        if len(specific_word) == 10:
            break
    non_stop_specific_for_each_category.append(specific_word)
    print(categories[c], ": ", specific_word)

business :  ['year', 'mr', 'growth', 'sales', 'economy', 'market', 'bank', 'firm', 'oil', 'new']
tech :  ['people', 'mobile', 'software', 'mr', 'phone', 'microsoft', 'music', 'games', 'net', 'users']
politics :  ['mr', 'labour', 'election', 'blair', 'party', 'brown', 'government', 'people', 'tax', 'howard']
sport :  ['england', 'game', 'year', 'world', 'win', 'time', 'wales', 'cup', 'ireland', 'chelsea']
entertainment :  ['film', 'best', 'year', 'music', 'band', 'number', 'awards', 'new', 'award', 'won']


Output of non-stop words list:
    
business :  ['year', 'mr', 'growth', 'sales', 'economy', 'market', 'bank', 'firm', 'oil', 'new']
    
tech :  ['people', 'mobile', 'software', 'mr', 'phone', 'microsoft', 'music', 'games', 'net', 'users']
    
politics :  ['mr', 'labour', 'election', 'blair', 'party', 'brown', 'government', 'people', 'tax', 'howard']
    
sport :  ['england', 'game', 'year', 'world', 'win', 'time', 'wales', 'cup', 'ireland', 'chelsea']
    
entertainment :  ['film', 'best', 'year', 'music', 'band', 'number', 'awards', 'new', 'award', 'won']

# Reimplementing naive bayes Algorithm for part 3

In [74]:
def improved_naive_bayes(unigrams,bigrams,test,specific_words,absence_words):
    test = np.ndarray.tolist(test)
    import re
    global number_of_unique_words, length_of_words_categorically
    for i in range(len(test)):
        test[i] = test[i][0].lower()
        test[i] = re.sub(r'[^a-zA-Z0-9]', ' ', test[i])
        test[i] = test[i].split()
    predictions_unigram = []
    predictions_bigram = []
    for i in range(len(test)):
        scores_unigram = [0,0,0,0,0]
        scores_bigram = [0,0,0,0,0]
        for k in range(5):
            for unique_word in set(test[i]):
                if unique_word in specific_words[k]:
                    # because when we look at the final scores of each text 
                    # for each category with bag of unigram
                    # difference between min score and max score is in general 
                    # about half of the length of text
                    # and in general every presence word is in text. So , this will work 10 times
                    #  so if an absence word of a class is in a text
                    # we are decreasing the probability of the class but we are not doing it impossible.
                    scores_unigram[k] += len(test[i]) * 0.05
                    # because when we look at the final scores of each text 
                    # for each category with bag of bigram
                    # difference between min score and max score is in general 
                    # about 7 times of the length of text
                    # and in general every presence word is in text. So , this will work 10 times
                    # so if an absence word of a class is in a text
                    # we are decreasing the probability of the class but we are not doing it impossible.
                    scores_bigram[k] += len(test[i]) * 0.35
            for j in range(len(test[i])):
                if test[i][j] in absence_words[k]:
                    # because when we look at the final scores of each text 
                    # for each category with bag of unigram
                    # difference between min score and max score is in general about half of
                    # the length of text so if an absence word of a class is in a text
                    # we are decreasing the probability of the class but we are not doing it impossible.
                    scores_unigram[k] -= len(test[i]) * 0.5
                    # because when we look at the final scores of each text 
                    # for each category with bag of bigram
                    # difference between min score and max score is in general 
                    # about 7 times of the length of text
                    # so if an absence word of a class is in a text
                    # we are decreasing the probability of the class but we are not doing it impossible.
                    scores_bigram[k] -= len(test[i]) * 3.5
                if test[i][j] in unigrams[categories[k]].keys():
                    scores_unigram[k] += np.log((unigrams[categories[k]][test[i][j]] +1) / (length_of_words_categorically[k] + number_of_unique_words))
                else:
                    scores_unigram[k] += np.log(1 / (length_of_words_categorically[k] + number_of_unique_words))
                if j+2 < len(test[i]):
                    bigram = test[i][j] + ' ' + test[i][j+1]
                    if bigram in bigrams[categories[k]].keys():
                        scores_bigram[k] += np.log((bigrams[categories[k]][bigram] + 1) / (length_of_words_categorically[k] + number_of_unique_words))
                    else:
                        scores_bigram[k] += np.log(1/(length_of_words_categorically[k]+number_of_unique_words))
        predictions_unigram.append(categories[scores_unigram.index(max(scores_unigram))])
        predictions_bigram.append(categories[scores_bigram.index(max(scores_bigram))])
    return predictions_unigram , predictions_bigram

predictions = naive_bayes(bag_of_unigrams,bag_of_bigrams,test)
print("Accuracy of part2 naive bayes unigram :" ,accuracy(test[:,1],predictions[0]))
print("Accuracy of part2 naive bayes bigram :" ,accuracy(test[:,1],predictions[1]))
predictions2 = improved_naive_bayes(bag_of_unigrams,bag_of_bigrams,test,[[],[],[],[],[]],absence_word_for_each_category)
print("Accuracy of part3 naive bayes unigram only using precences :" ,accuracy(test[:,1],predictions2[0]))
print("Accuracy of part3 naive bayes bigram only using precences :" ,accuracy(test[:,1],predictions2[1]))
predictions3 = improved_naive_bayes(bag_of_unigrams,bag_of_bigrams,test,specific_word_for_each_category,[[],[],[],[],[]])
print("Accuracy of part3 naive bayes unigram only using absences :" ,accuracy(test[:,1],predictions3[0]))
print("Accuracy of part3 naive bayes bigram only using absences :" ,accuracy(test[:,1],predictions3[1]))
predictions4 = improved_naive_bayes(bag_of_unigrams,bag_of_bigrams,test,specific_word_for_each_category,absence_word_for_each_category)
print("Accuracy of part3 naive bayes unigram using both absences and presences :" ,accuracy(test[:,1],predictions4[0]))
print("Accuracy of part3 naive bayes bigram using both absences and presences :" ,accuracy(test[:,1],predictions4[1]))
predictions5 = improved_naive_bayes(bag_of_unigrams,bag_of_bigrams,test,non_stop_specific_for_each_category,[[],[],[],[],[]])
print("Accuracy of part3 naive bayes unigram only using nonstop presences :" ,accuracy(test[:,1],predictions5[0]))
print("Accuracy of part3 naive bayes bigram only using nonstop presences :" ,accuracy(test[:,1],predictions5[1]))


Accuracy of part2 naive bayes unigram : 99.32885906040268
Accuracy of part2 naive bayes bigram : 100.0
Accuracy of part3 naive bayes unigram only using precences : 98.65771812080537
Accuracy of part3 naive bayes bigram only using precences : 98.99328859060402
Accuracy of part3 naive bayes unigram only using absences : 99.32885906040268
Accuracy of part3 naive bayes bigram only using absences : 98.32214765100672
Accuracy of part3 naive bayes unigram using both absences and presences : 98.65771812080537
Accuracy of part3 naive bayes bigram using both absences and presences : 97.31543624161074
Accuracy of part3 naive bayes unigram only using nonstop presences : 99.32885906040268
Accuracy of part3 naive bayes bigram only using nonstop presences : 94.63087248322148


# Output 1:

Accuracy of part2 naive bayes unigram : 98.99328859060402

Accuracy of part2 naive bayes bigram : 99.66442953020135

Accuracy of part3 naive bayes unigram only using precences : 98.99328859060402

Accuracy of part3 naive bayes bigram only using precences : 99.32885906040268

Accuracy of part3 naive bayes unigram only using absences : 98.99328859060402

Accuracy of part3 naive bayes bigram only using absences : 96.97986577181209

Accuracy of part3 naive bayes unigram using both absences and presences : 98.99328859060402

Accuracy of part3 naive bayes bigram using both absences and presences : 96.64429530201342

Accuracy of part3 naive bayes unigram only using nonstop presences : 98.65771812080537

Accuracy of part3 naive bayes bigram only using nonstop presences : 95.63758389261746

# Output 2:

Accuracy of part2 naive bayes unigram : 100.0

Accuracy of part2 naive bayes bigram : 100.0

Accuracy of part3 naive bayes unigram only using precences : 100.0

Accuracy of part3 naive bayes bigram only using precences : 100.0

Accuracy of part3 naive bayes unigram only using absences : 100.0

Accuracy of part3 naive bayes bigram only using absences : 97.31543624161074

Accuracy of part3 naive bayes unigram using both absences and presences : 100.0

Accuracy of part3 naive bayes bigram using both absences and presences : 97.31543624161074

Accuracy of part3 naive bayes unigram only using nonstop presences : 99.66442953020135

Accuracy of part3 naive bayes bigram only using nonstop presences : 95.30201342281879

# Output 3

Accuracy of part2 naive bayes unigram : 99.32885906040268

Accuracy of part2 naive bayes bigram : 100.0

Accuracy of part3 naive bayes unigram only using precences : 98.65771812080537

Accuracy of part3 naive bayes bigram only using precences : 98.99328859060402

Accuracy of part3 naive bayes unigram only using absences : 99.32885906040268

Accuracy of part3 naive bayes bigram only using absences : 98.32214765100672

Accuracy of part3 naive bayes unigram using both absences and presences : 98.65771812080537

Accuracy of part3 naive bayes bigram using both absences and presences : 97.31543624161074

Accuracy of part3 naive bayes unigram only using nonstop presences : 99.32885906040268

Accuracy of part3 naive bayes bigram only using nonstop presences : 94.63087248322148

# Effect of removing stop-words when improving model

In [78]:
vectorizer = CountVectorizer(stop_words=ENGLISH_STOP_WORDS) # CountVectorizer for unigram matrix.
vectorizer2 = CountVectorizer(ngram_range=(2, 2),stop_words=ENGLISH_STOP_WORDS)# CountVectorizer for bigram matrix.
whole_matrix = vectorizer.fit_transform(training[:, 0]).toarray()
number_of_unique_words = whole_matrix.shape[1]
whole_matrix_of_bigrams = vectorizer2.fit_transform(training[:, 0]).toarray()
number_of_unique_bigrams = whole_matrix_of_bigrams.shape[1]
bag_of_unigrams = defaultdict()
bag_of_bigrams = defaultdict()
# according to unigram and bigram matrices, creating dictionaries respectively
for c in range(len(whole_category)):
    cat_name = whole_category[c]
    X = vectorizer.fit_transform(cat_name)
    unigram_matrix = X.toarray()
    words = (vectorizer.get_feature_names_out())
    bigram_matrix = vectorizer2.fit_transform(cat_name).toarray()
    bigrams = vectorizer2.get_feature_names_out()
    vocabulary = defaultdict(int) # vocabulary of unigrams for each category
    vocabulary2 = defaultdict(int) # vocabulary of bigrams for each category
    sumOfColumns = np.sum(unigram_matrix, axis=0)
    # adding existing words to unigram vocabulary
    for i in range(unigram_matrix.shape[0]):
        for j in range(unigram_matrix.shape[1]):
            if unigram_matrix[i][j] == 0:
                continue
            else:
                vocabulary[words[j]] += unigram_matrix[i][j]
    # collecting unigram vocabularies for each category
    bag_of_unigrams[categories[c]] = vocabulary
    # adding existing words to bigram vocabulary
    for i in range(bigram_matrix.shape[0]):
        for j in range(bigram_matrix.shape[1]):
            if bigram_matrix[i][j] == 0:
                continue
            else:
                vocabulary2[bigrams[j]] += bigram_matrix[i][j]
    # collecting bigram vocabularies for each category
    bag_of_bigrams[categories[c]] = vocabulary2
    
def naive_bayes(unigrams,bigrams,test):
    test = np.ndarray.tolist(test)
    import re
    global number_of_unique_words, length_of_words_categorically
    for i in range(len(test)):
        test[i] = test[i][0].lower() # converting every letter to lowercase 
        test[i] = re.sub(r'[^a-zA-Z0-9]', ' ', test[i]) # we are removing punctuations for not considering them as a word
        test[i] = test[i].split() # converting text string to string list of every word.
    predictions_unigram = []
    predictions_bigram = []
    for i in range(len(test)):
        scores_unigram = [0,0,0,0,0] # keeping scores of each category for unigram
        scores_bigram = [0,0,0,0,0] # keeping scores of each category for bigram
        for j in range(len(test[i])):
            for k in range(5):
                if test[i][j] in unigrams[categories[k]].keys():
                    scores_unigram[k] += np.log((unigrams[categories[k]][test[i][j]] +1) / (length_of_words_categorically[k] + number_of_unique_words))
                else:
                    scores_unigram[k] += np.log(1 / (length_of_words_categorically[k] + number_of_unique_words))
                if j+2 < len(test[i]):
                    bigram = test[i][j] + ' ' + test[i][j+1]
                    if bigram in bigrams[categories[k]].keys():
                        scores_bigram[k] += np.log((bigrams[categories[k]][bigram] + 1) / (length_of_words_categorically[k] + number_of_unique_words))
                    else:
                        scores_bigram[k] += np.log(1/(length_of_words_categorically[k]+number_of_unique_words))

        predictions_unigram.append(categories[scores_unigram.index(max(scores_unigram))])
        predictions_bigram.append(categories[scores_bigram.index(max(scores_bigram))])
    return predictions_unigram , predictions_bigram

def accuracy(outputs,predictions):
    correct_predictions = 0
    for i in range(len(outputs)):
        if outputs[i] == predictions[i]:
            correct_predictions +=1
    return 100 * correct_predictions / len(outputs)

predictions = naive_bayes(bag_of_unigrams,bag_of_bigrams,test)
print("Unigram: ",accuracy(test[:,1],predictions[0]))
print("Bigram: ",accuracy(test[:,1],predictions[1]))



Unigram:  99.66442953020135
Bigram:  65.43624161073825


Output 1:

Unigram:  99.66442953020135
    
Bigram:  65.43624161073825
    
Output 2:

Unigram:  99.66442953020135
    
Bigram:  69.79865771812081
    
Output 3:

Unigram:  99.32885906040268
    
Bigram:  71.81208053691275

The reason behind decreasing accuracy rates of bigram is that when we remove stopwords 
from the entire dataset, unrelated word pairs will increase.

# Part 4 accuracy algorithm

In [ ]:
def accuracy(outputs,predictions):
    correct_predictions = 0
    for i in range(len(outputs)):
        if outputs[i] == predictions[i]:
            correct_predictions +=1
    return 100 * correct_predictions / len(outputs)